app/controllers/provider_interface/provider_users_controller.rb

```diff
@@ -33,15 +39,42 @@ def create
       redirect_to provider_interface_provider_users_path
     end
 
+    def edit_providers
+      provider_user = ProviderUser.find(params[:provider_user_id])
+      @form = ProviderUserForm.from_provider_user(provider_user)
+      @form.current_provider_user = current_provider_user
+    end
+
+    def update_providers
+      provider_user = ProviderUser.find(params[:provider_user_id])
+      provider_user.assign_attributes(provider_user_params.except(:permissions))
+      @form = ProviderUserForm.from_provider_user(provider_user)
+      service = SaveProviderUser.new(provider_user: provider_user, permissions: permissions_params)
```

Comment:
If the user we're editing has access to provider A and B, and the `current_provider_user` only has access to A, when the form is submitted, the user's access to B is accidentally revoked.

This is what I tried to warn about in https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/1889#discussion_r409808564. Do we need some kind of extra layer to avoid us accidentally doing this? It's a pretty tricky thing to keep in mind.

Reply:
Ah yes, we should only unassign what is present in the form and unchecked then. :+1:

Reply:
@tijmenb https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/pull/1931/commits/689a2b97504f82f7225cd55cd4d093836f12991b attempts to address this. I think we'll need to perform a similar operation for permissions as the same problem exists, will see if I can produce something common to both.

Reply:
Managing a provider association is the same thing as managing a permission. If the form only had the power to act on a certain set of permissions then could we make the updates much dumber?

Reply:
@duncanjbrown I see what you mean. Handling all this in the service layer would mean doing the updates in a single transaction. @willmcb and I have been pairing on a way to do this.

Reply:
@steventux I did some sketching. Got a bit caught up with Rails form magic, but I think this approach might be worth a go.

commit 1, which pilots it on the `show` action: cb9e1029d225518bb95ed787b7cb904a4ebd7047

commit 2, which is an (incomplete) attempt to approach editing in a similar way
361e2743427f6ddba9bb1f4b30338ffd37b88bca

If this worked I think we could get rid of the intermediate representation of permissions in the OptionsService and various other things here and in support

Reply:
slept on it and came up with https://github.com/DFE-Digital/apply-for-postgraduate-teacher-training/commit/fc8e646adf15d0b11f6b8f48f6987dd3d25be3ac which I think could work